russound_rio/rio.py

```diff
@@ -387,15 +468,57 @@ def unwatch_source(self, source_id):
                 self._send_cmd("WATCH S[%d] OFF" % (
                     source_id, )))
 
+    @asyncio.coroutine
+    def get_preset_variable(self, preset_id, variable):
+        """ Retrieve the current value of a preset variable.  If the variable is
+        not found in the local cache then the value is requested from the
+        controller.  """
+
+        try:
+            return self._retrieve_cached_preset_variable(preset_id, variable)
+        except UncachedVariable:
+            return (yield from self._send_cmd("GET %s.%s" % (
+                preset_id.device_str(), variable)))
+
+    @asyncio.coroutine
+    def calc_preset_index(self, bank_id, preset_id):
+        """
+        Calculates the index of a preset.
+        """
+        return (((bank_id - 1) * 2) + preset_id)
+        
     @asyncio.coroutine
     def enumerate_sources(self):
-        """ Return a list of (source_id, source_name) tuples """
+        """ Return a list of (source_id, source_name, source_type) tuples """
         sources = []
         for source_id in range(1, 17):
             try:
-                name = yield from self.get_source_variable(source_id, 'name')
-                if name:
-                    sources.append((source_id, name))
+                source_name = yield from self.get_source_variable(source_id, 'name')
+                source_type = yield from self.get_source_variable(source_id, 'type')
+                if source_name and source_type:
```

Comment:
May require tweaking if other Russound devices do not contain a source type. Although with MCA series this prevents unused sources from showing up.